STM32: Replace HAL register operations

gautierg-st opened the pull request:
This PR replaces the STM32 HAL register operations macros (such as WRITE_REG or MODIFY_REG) and functions (such as LL_I2C_WriteReg) by Zephyr equivalents in all STM32 drivers and SoC.

etienne-lms reviewed (commented):
I wonder if all the casts the `mem_addr_t` could be hidden some way with helper macros or inline functions. Outside of the scope of this P-R.

mathieuchopstm reviewed (commented):
Seems OK overall. I'm not a fan of using `sys_{read,write}N` when we have pointers already available, but I guess it may be useful to Cortex-A... (has it even been a problem though?)

Unrelated remark: the Cortex-A/R implementation of the helpers use `dmb()` as barrier but [Linux seems to use `dsb()`](https://github.com/torvalds/linux/blob/5472d60c129f75282d94ae5ad072ee6dfb7c7246/arch/arm/include/asm/barrier.h#L65).

etienne-lms commented on drivers/counter/counter_ll_stm32_rtc.c:
```diff
@@ -270,46 +270,46 @@ static int rtc_stm32_deinit(void)
-	WRITE_REG(RTC->PRLL, 0x8000U);
-	WRITE_REG(RTC->CRH, 0U);
-	WRITE_REG(RTC->CRL, 0x20U);
+	sys_write32(0U, (mem_addr_t)&RTC->CNTL);
```
Regarding the 2nd bullet, I would also be in favor of a STM32 macro since it's used always with pointers needed to be cast to `mem_addr_t`, e.g.:
```c
#define STM32_IOREG_READ32(base, field)			sys_read32((mem_addr_t)&(base)->field)
#define STM32_IOREG_WRITE32(base, field, value)		sys_write32((value), (mem_addr_t)&(base)->field)

static inline void stm32_io_set_bit32(void *ptr, uint32_t mask)
{
	sys_set_bits((mem_addr_t)ptr, mask); 
}
static inline void stm32_io_clear_bit32(void *ptr, uint32_t mask)
{
	sys_clear_bits((mem_addr_t)ptr, mask);
}
static inline void stm32_io_modify_bit32(void *ptr, uint32_t mask, uint32_t value)
{
	sys_modify_bits((mem_addr_t)ptr, mask, value);
}

static inline bool stm32_io_bitmask_all_set(void *ptr, uint32_t mask)
{
	return sys_test_bits((mem_addr_t)ptr, mask) == mask;
}

peter-mitsis commented on include/zephyr/sys/sys_io.h:
```diff
@@ -265,6 +265,18 @@ typedef uintptr_t mem_addr_t;
+ * @brief Modify the bits from addr to the values defined by @p mask and @p value
+ *
+ * This functions modifies the designated bits from addr by clearing mask before setting value.
+ * It is equivalent to calling sys_clear_bits then sys_set_bits, but addr is modified only once.
```
I would suggest rewording this to something like the following to better describe what it does, not how it does it ...

This function modifies the subset of bits defined by @p mask at address @p addr
to match the corresponding subset of bits in @p value. It is equivalent to calling
sys_clear_bits() and then sys_set_bits(), except that the data at the specified
address is only modified once.

gautierg-st committed:
soc: st: stm32: common: add common bitops functions

Add a set of bitops functions in order to replace the STM32 HAL bitops
macros throughout the drivers.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: stm32: replace SET_BIT HAL macro by stm32_reg_set_bits

For all STM32 drivers, replace the SET_BIT macro (defined in
the STM32 HAL) by stm32_reg_set_bits defined in Zephyr.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: stm32: replace CLEAR_BIT HAL macro by stm32_reg_clear_bits

For all STM32 drivers, replace the CLEAR_BIT macro (defined in
the STM32 HAL) by stm32_reg_clear_bits defined in Zephyr.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: stm32: replace READ_BIT HAL macro by stm32_reg_read_bits

For all STM32 drivers and SoC, replace the READ_BIT macro (defined in
the STM32 HAL) by stm32_reg_read_bits.
Fixes some cases where the return value was tested like a boolean
despite not being one.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: stm32: replace WRITE_REG HAL macro by stm32_reg_write

For all STM32 drivers and SoC, replace the WRITE_REG macro and the
LL_xxx_WriteReg functions (defined in the STM32 HAL) by
stm32_reg_write defined in Zephyr.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: stm32: replace READ_REG HAL macro by stm32_reg_read

For all STM32 drivers and SoC, replace the READ_REG macro and the
LL_xxx_ReadReg functions (defined in the STM32 HAL) by
stm32_reg_read defined in Zephyr.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: stm32: replace MODIFY_REG HAL macro by stm32_reg_modify_bits

For all STM32 drivers and SoC, replace the MODIFY_REG macro (defined in
the STM32 HAL) by stm32_reg_modify_bits defined in Zephyr.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mathieuchopstm commented on soc/st/stm32/common/stm32_bitops.h:
```diff
@@ -0,0 +1,50 @@
+
+static ALWAYS_INLINE uint32_t stm32_read_mask(volatile uint32_t *addr, uint32_t mask)
+{
+	return (*addr) & mask;
```
Use the `sys_read32` primitive:
```suggestion
	return sys_read32((mem_addr_t)addr) & mask;
```

gautierg-st replied:
Done

gautierg-st commented on soc/st/stm32/common/stm32_bitops.h:
```diff
@@ -0,0 +1,50 @@
+#include <zephyr/arch/common/sys_bitops.h>
+#include <zephyr/arch/common/sys_io.h>
+
+static ALWAYS_INLINE void stm32_set_mask(volatile uint32_t *addr, uint32_t mask)
```
I chose `stm32_reg_write / stm32_reg_read / stm32_reg_set_bits / stm32_reg_clear_bits / stm32_reg_read_bits / stm32_reg_modify_bits`

etienne-lms commented on soc/st/stm32/common/stm32_bitops.h:
```diff
@@ -0,0 +1,50 @@
+}
+
+static ALWAYS_INLINE void stm32_reg_modify_bits(volatile uint32_t *addr, uint32_t mask,
+						uint32_t value)
```
Could fit in a single line.

etienne-lms commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -467,7 +468,8 @@ static struct flash_stm32_sector_t get_sector(const struct device *dev, off_t of
 	/* Check whether bank1/2 are swapped */
-	bank_swap = (READ_BIT(FLASH->OPTCR, FLASH_OPTCR_SWAP_BANK) == FLASH_OPTCR_SWAP_BANK);
+	bank_swap = stm32_reg_read_bits(&FLASH->OPTCR, FLASH_OPTCR_SWAP_BANK) ==
+		    FLASH_OPTCR_SWAP_BANK;
```
Proposal to fix the missing empty line between variable definition and the instructions that follows:
```suggestion
	/* Check whether bank1/2 are swapped */
	bool bank_swap = stm32_reg_read_bits(&FLASH->OPTCR, FLASH_OPTCR_SWAP_BANK) ==
			 FLASH_OPTCR_SWAP_BANK;
```
feel from to discard this review comment.

etienne-lms commented on drivers/usb_c/tcpc/ucpd_stm32.c:
```diff
@@ -1110,9 +1110,9 @@ static void ucpd_isr(const struct device *dev_inst[])
+	stm32_reg_read(&((const struct tcpc_config *)dev_inst[0]->config)->ucpd_port->SR);
 	sr1 =
-	LL_UCPD_ReadReg(((const struct tcpc_config *)dev_inst[1]->config)->ucpd_port, SR);
+	stm32_reg_read(&((const struct tcpc_config *)dev_inst[1]->config)->ucpd_port->SR);
```
These fit each in a single line.

etienne-lms commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -224,7 +224,7 @@ static void configure_rng(void)
 
 	if (cur_nist_cfg != desired_nist_cfg || cur_htcr != desired_htcr) {
-		MODIFY_REG(rng->CR, cur_nist_cfg, (desired_nist_cfg | RNG_CR_CONDRST));
+		stm32_reg_modify_bits(&rng->CR, cur_nist_cfg, (desired_nist_cfg | RNG_CR_CONDRST));
```
By the way, drop the useless parentheses?
```suggestion
		stm32_reg_modify_bits(&rng->CR, cur_nist_cfg, desired_nist_cfg | RNG_CR_CONDRST);

gautierg-st committed:
drivers: i2c: stm32: v2: simplify reg operation

Call stm32_reg_set_bits instead of stm32_reg_read then stm32_reg_write.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

etienne-lms commented on drivers/adc/adc_stm32.c:
```diff
@@ -868,9 +869,9 @@ static void set_reg_value(const struct device *dev, uint32_t reg,
 	ADC_TypeDef *adc = config->base;
 
-	uintptr_t addr = (uintptr_t)adc + reg;
+	uint32_t *addr = (uint32_t *)adc + reg;
```
To be fixed: the passed `reg` value is a byte offset. Here it becomes a 32bit word offset.
Suggestion:
```c
	const struct adc_stm32_cfg *config = dev->config;
	size_t reg32_offset = reg / sizeof(uint32_t);
	uint32_t *addr = (uint32_t *)config->base + reg32_offset;

etienne-lms reviewed (changes requested):
A few review comments that may be addressed later but an issue to fix in *drivers/adc/adc_stm32.c* `set_reg_value()`.

gautierg-st committed:
drivers: usb_c: tcpc: stm32: use dedicated bitops functions

Use dedicated set/clear bits function instead of reading/masking/writing
manually.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented on drivers/usb_c/tcpc/ucpd_stm32.c:
```diff
@@ -585,16 +583,15 @@ static void ucpd_start_transmit(const struct device *dev,
 		imr = LL_UCPD_ReadReg(config->ucpd_port, IMR);
 		imr |= UCPD_IMR_HRSTDISCIE | UCPD_IMR_HRSTSENTIE;
-		LL_UCPD_WriteReg(config->ucpd_port, IMR, imr);
+		stm32_reg_write(&config->ucpd_port->IMR, imr);
```
Done in specific commit

gautierg-st commented:
Rebased to fix conflicts.

etienne-lms reviewed (dismissed):
LAGTM

etienne-lms reviewed (approved):
Still LGTM.

mathieuchopstm commented on drivers/adc/adc_stm32.c:
```diff
@@ -865,11 +867,10 @@ static void set_reg_value(const struct device *dev, uint32_t reg,
-
-	uintptr_t addr = (uintptr_t)adc + reg;
+	size_t reg32_offset = reg / sizeof(uint32_t);
+	uint32_t *addr = (uint32_t *)config->base + reg32_offset;
```
> > This should have raised a compiler warning (I didn't check if it did or not)
> 
> The many places where `stm32_reg_xxx()` are used use a pointer argument that is not an explicit `volatile` data pointer. I don't think it's needed here. Actaully, I don't think the `address` argument of these helper functions need to be `volatile`. The cast is done in the sys_xxx() functions and that should be enough.

You're right, it is probably allowed to up-cast a `uint32_t *` to `volatile uint32_t *` because the former is less strict than the latter (same as `const uint32_t` -> `const uint32_t *`). Still better to cast to proper type explicitly IMO.
